auth/jwt/jwt_test.go

```diff
@@ -40,288 +38,228 @@ func TestJWTCaching(t *testing.T) {
 		if handler, ok := keyForPath[r.URL.Path]; ok && !called[r.URL.Path] {
 			called[r.URL.Path] = true
 			handler(w, r)
 		} else {
-			send401Handler(w, r)
+			authtest.DeniedHandler()(w, r)
 		}
 	}))
 	defer ts.Close()
 
-	// Refresh time is too small and will be overriden.
-	hasProvider := Provider{JWKSURL: ts.URL + "/hasit", Refresh: 10 * time.Second}
-	missingProvider := Provider{JWKSURL: ts.URL + "/doesnothaveit", Refresh: 10 * time.Second}
+	goodProvider := Provider{JWKSURL: ts.URL + "/hasit", Refresh: -time.Second}
+	goodProvider2 := Provider{JWKSURL: ts.URL + "/hasit"}
+	wrongProvider := Provider{JWKSURL: ts.URL + "/doesnothaveit"}
+	badUrlProvider := Provider{JWKSURL: "badurl"}
+	badFormatProvider := Provider{JWKSURL: ts.URL + "/badformat"}
+	deniedProvider := Provider{JWKSURL: ts.URL + "/denied"}
 	jwtVerifier := NewVerifier(VerifierOptions{
-		Providers: []Provider{missingProvider, hasProvider},
+		Client:    http.DefaultClient,
+		Providers: []Provider{goodProvider, goodProvider2, wrongProvider, badUrlProvider, badFormatProvider, deniedProvider},
 	})
 	jwtVerifier.Start()
+	time.Sleep(10 * time.Millisecond)
 	defer jwtVerifier.Stop()
 
-	// Make sure the missing provider caches its results first.
-	_, err = jwtVerifier.Parse(jwt, missingProvider)
-	if err == nil {
-		t.Errorf("no error found checking %q, expected error", missingProvider.JWKSURL)
-	}
-
-	for i := 0; i < 5; i++ {
-		// Do a first request and confirm that things look good.
-		_, err = jwtVerifier.Parse(jwt, hasProvider)
-		if err != nil {
-			t.Fatal(err)
-		}
-		_, err = jwtVerifier.Parse(noExpireJwt, hasProvider)
-		if err != nil {
-			t.Fatal(err)
-		}
-	}
-
-	// Ensure that good results are only cached on the correct provider.
-	_, err = jwtVerifier.Parse(jwt, missingProvider)
-	if err == nil {
-		t.Errorf("no error found checking %q, expected error", missingProvider.JWKSURL)
+	for _, test := range []struct {
+		desc           string
+		provider       Provider
+		wantParseError bool
+		wantJwksError  bool
+	}{
+		{"good provider", goodProvider, false, false},
+		{"wrong provider", wrongProvider, true, false},
+		{"bad provider url", badUrlProvider, true, true},
+		{"bad jwks format", badFormatProvider, true, true},
+		{"denied provider", deniedProvider, true, true},
+	} {
+		t.Run("parse "+test.desc, func(t *testing.T) {
+
+			// check first run & cached run
+			for i := 0; i < 2; i++ {
+				_, err = jwtVerifier.Parse(jwt, test.provider)
+				if test.wantParseError && err == nil {
+					t.Error("wanted error, got none")
+				}
+				if !test.wantParseError && err != nil {
+					t.Errorf("wanted no error, got: %v", err)
+				}
+			}
+		})
+
+		t.Run("caches "+test.desc, func(t *testing.T) {
+
+			cacheKey := fmt.Sprintf("%s-%s", test.provider.JWKSURL, jwt)
+			v, ok := jwtVerifier.(*verifier).cache.Get(cacheKey)
+			if test.wantParseError && ok {
+				t.Errorf("want nothing in verifier cache, got: %v", v)
```

Comment:
It'd be simpler to merge two conditions into `if test.wantParseError == ok`. The error msg could be something like `(got: %v in verifier cache, wanted something = %v, v, ok)`